src/NuGet.Core/NuGet.Protocol/Converters/V3SearchResultsConverter.cs

```diff
@@ -0,0 +1,106 @@
+using Newtonsoft.Json;
+using NuGet.Protocol.Model;
+using System;
+
+namespace NuGet.Protocol.Converters
+{
+    public class V3SearchResultsConverter : JsonConverter
+    {
+        public int Take { get; set; }
+
+        public V3SearchResultsConverter(int take)
+        {
+            Take = take;
+        }
+
+        public override bool CanWrite => false;
+
+        public override bool CanConvert(Type objectType) => objectType == typeof(V3SearchResults);
+
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        {
+            if (objectType != typeof(V3SearchResults))
+            {
+                throw new InvalidOperationException();
+            }
+
+            if (reader.TokenType != JsonToken.StartObject)
+            {
+                throw new JsonException("Expected StartObject, found " + reader.TokenType);
+            }
+
+            var searchResults = new V3SearchResults();
+
+            if (Take <= 0)
+            {
+                return searchResults;
+            }
+
+            var finished = false;
+
+            while (!finished)
+            {
+                reader.Read();
+
+                switch (reader.TokenType)
+                {
+                    case JsonToken.PropertyName:
+                        switch ((string)reader.Value)
+                        {
+                            case "totalHits":
+                                long totalHits = long.TryParse(reader.ReadAsString(), out totalHits) ? totalHits : 0;
+                                searchResults.TotalHits = totalHits;
+                                break;
+
+                            case "data":
+                                reader.Read();
+
+                                if (reader.TokenType != JsonToken.StartArray)
+                                {
+                                    throw new JsonException("data should be an array");
+                                }
+
+                                reader.Read();
+
+                                while (reader.TokenType != JsonToken.EndArray)
+                                {
+                                    var searchResult = JsonExtensions.JsonObjectSerializer.Deserialize<PackageSearchMetadata>(reader);
+
+                                    searchResults.Data.Add(searchResult);
+                                    Take--;
+
+                                    if (Take <= 0)
+                                    {
+                                        finished = true;
```

Comment:
If we don't process item while iterating then we have to either add things to some buffer array for alter processing which will increase our memory footprint or have to re-do reading from the beginning. I'm not sure if it's desired thing, so I'll wait for input from other team members.

Reply:
Why was this conversation marked as resolved? No one else commented.

@nkolev92 what do you think?  We have to trade off correctness for performance.


@erdembayar I don't fully understand your comment:

> If we don't process item while iterating then we have to either add things to some buffer array for later processing which will increase our memory footprint or have to re-do reading from the beginning

no, we don't have to add anything to any buffer for later processing, we either:

```cs
var searchResult = serializer.Deserialize<PackageSearchMetadata>(reader);
if (searchResults.Data.Count < _take)
{
  searchResults.Data.Add(searchResult);
}
```

or 
```cs
if (searchResults.Data.Count < _take)
{
  var searchResult = serializer.Deserialize<PackageSearchMetadata>(reader);
  searchResults.Data.Add(searchResult);
}
else
{
  serializer.Skip();
}
```
This way, we let the serializer check the syntax, but we don't store anything, so we don't use any more memory (the first option will have higher GC)

Reply:
@zivkan but in case this static package source it'll will return thousands items instead of 25 but it still wouldn't cause any syntax exception, since it's valid json result. Now we're doing additional work of going through thousands items.
@nkolev92 please give your input?

Reply:
> but in case this static package source it'll return thousands items instead of 25

The method returns `searchResults`. If we only add 25 items to `searchResults.Data`, then it will return 25 items, even if the JsonReader/JsonSerializer has checked the syntax of thousands.

Reply:
@zivkan Yes, I understand it still returns correct 25 items in the end. But before that it'll just go through thousands of items which don't need for end result (since it's correct json it won't through any exception during that time).

Reply:
How big of a difference are we talking about here, can we profile the performance of both if it's easy?

I think both arguments have some merit, but it'd be easier to understand what we're trading off here.

Reply:
@zivkan  @nkolev92 I did benchmarking on 2 scenarios.
On small load json (184kb) both process time and memory foot print was exact same as expected.
But on extra large json load (around 145mb) reading to end of json was 2.27 times slow and memory footprint was 1.41 times larger. Of cause such large json input is kind of edge case.

![image](https://user-images.githubusercontent.com/8766776/83099872-55ce2080-a063-11ea-874f-f477064a9c1c.png)

Please advice.

Reply:
@zivkan 
Yes, I also believe it's something to do with bytes are converted to strings for Newtonsoft.json consumption. If we really aim for performance try working raw bytes but maintaining code will be nightmare. 
Can I mark this comment as resolved now?

Reply:
>  If we really aim for performance try working raw bytes but maintaining code will be nightmare.

Yeah, as you said, we should not do that :) 

Parsing efficiently is never trivial and it's better to rely on true and tried libraries for that. 

At some point whenever we can switch System.Text.Json will likely will perform better in this scenario. 

We should take the wins we have :)

Reply:
Just looping back here, I think it's ok to do the early exit. 
We're really optimizing for the sleet case here, so we should take every reasonable win.